x/gamm/keeper/pool.go

```diff
@@ -3,16 +3,61 @@ package keeper
 import (
 	"fmt"
 
-	gogotypes "github.com/gogo/protobuf/types"
-
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	gogotypes "github.com/gogo/protobuf/types"
 
 	"github.com/osmosis-labs/osmosis/v12/osmoutils"
 	"github.com/osmosis-labs/osmosis/v12/x/gamm/pool-models/balancer"
 	"github.com/osmosis-labs/osmosis/v12/x/gamm/types"
 )
 
+// TODO spec and tests
+func (k Keeper) InitializePool(ctx sdk.Context, pool types.PoolI, creatorAddress sdk.AccAddress) error {
+	poolId := pool.GetId()
+
+	// Finally, add the share token's meta data to the bank keeper.
+	poolShareBaseDenom := types.GetPoolShareDenom(poolId)
+	poolShareDisplayDenom := fmt.Sprintf("GAMM-%d", poolId)
+	k.bankKeeper.SetDenomMetaData(ctx, banktypes.Metadata{
+		Description: fmt.Sprintf("The share token of the gamm pool %d", poolId),
+		DenomUnits: []*banktypes.DenomUnit{
+			{
+				Denom:    poolShareBaseDenom,
+				Exponent: 0,
+				Aliases: []string{
+					"attopoolshare",
+				},
+			},
+			{
+				Denom:    poolShareDisplayDenom,
+				Exponent: types.OneShareExponent,
+				Aliases:  nil,
+			},
+		},
+		Base:    poolShareBaseDenom,
+		Display: poolShareDisplayDenom,
+	})
+
+	// Mint the initial pool shares share token to the sender
+	err := k.MintPoolShareToAccount(ctx, pool, creatorAddress, pool.GetTotalShares())
+	if err != nil {
+		return err
+	}
+
+	traditionalPool, ok := pool.(types.TraditionalAmmInterface)
+	if !ok {
+		return fmt.Errorf("failed to create gamm pool. c+Could not cast to TraditionalAmmInterface")
+	}
+
```

Comment:
can we move the check to the very top so that we don't do unnecessary operations before failing